### PR TITLE
[MFMA] Refactor dot pipeline to reduce code duplication

### PR DESCRIPTION
**Pull request by binarman**
This PR:
- simplifies data types generated by `shared->mfma dot op` layout conversions. Do not pack data types in int32 or int64
- reduce code duplication between fast/normal path
- reduce code duplication between operand A and operand B

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 1]
```diff
@@ -41,16 +41,14 @@ Type getShemPtrTy(Type elemTy) {
...
-  Value waveMN = udiv(wave, i32_val(wpt[0]));
-  return urem(urem(waveMN, i32_val(wpt[1])), i32_val(N / elemPerInstr));
+// Get waveId inside block of waves.
+Value getWaveIdInBlock(ConversionPatternRewriter &rewriter, Location loc,
+                       Value waveId, const ArrayRef<unsigned int> &wpt,
+                       int elemPerInstrNonK, int tensorSizeNonK, int nonKIdx) {
+  if (nonKIdx == 1)
+    waveId = udiv(waveId, i32_val(wpt[0]));
+  return urem(urem(waveId, i32_val(wpt[nonKIdx])),
+              i32_val(tensorSizeNonK / elemPerInstrNonK));
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:51]: I'm confused about this part.
Say we have warpsPerCTA={2,2}, waveId=1, then we are talking about the top left wave in the workgroup, right?
If so, we have
- for opA, i.e. nonKIdx=0, spatialWarpId = (waveId % wpt[0]) % (M / 32) = 1
- for opB, i.e. nonKIdx=1, spatialWarpId = ((waveId/wpt[0])%wpt[1]) % (N / 8) = 0

But shouldn't wave1 has index 1 for opB and index 0 for opA?
example

File: lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/MFMA.cpp [hunk 2]
```diff
@@ -331,13 +334,29 @@ struct DotOpMFMAConversionHelper {
...
                                                  Type type) const {
     auto elems = typeConverter->unpackLLElements(loc, value, rewriter, type);
     ValueTable vals;
     for (int i = 0; i < n0; i++) {
       for (int j = 0; j < n1; j++) {
-        vals[{i, j}] = elems[n1 * i + j];
+        auto rawElems = elems[n1 * i + j];
+        Value convertedElems;
+        if (type.isF32()) {
+          convertedElems = extract_element(type, rawElems, i32_val(0));
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/MFMA.cpp:349]: Why do we need to dereference rawElems here for i32 input?

Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/MFMA.cpp:349]: This part is related to processing of A/B dot operands.
This code is an adapter from generic `vec<base_type x kwidth>` format to the format that rocdl intrinsic expects.

Previously this transformation was done in Type converter (see [TypeConverter.cpp](https://github.com/ROCmSoftwarePlatform/triton/pull/400/files#diff-928f898f93db4afa3551c43a4bdc36b127c0c866f76ac7cdc90bae10116972e9) below), but I feel that it is better to have this transformation closer to MFMA emitting code. So I simplified type converter and moved this code here.

A and B could be one of variety of types: fp32, fp16, bf16, int8, fp8*
But rocdl mfma intrinsics takes some of these types in a packed integer format.

For example:
- `fp32xfp32 -> fp32` version intrinsic takes plain scalar `fp32`, `fp32` as A/B arguments.
- `fp8xfp8 -> fp32` or `int8xint8 -> fp32` version takes several values in for of packed int32 or int64 (depending on kwidth of operation)

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp (hunk 1 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:51]: > But shouldn't wave1 has index 1 for opB and index 0 for opA?

It was originally implemented like this. I am not sure, if one or another orientation have advantages. I did not try other layout.

If you think transposed wave indexing could have advantages or preferable for style reasons, we can swap it and see what happens.

Comment (zhanglx13) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:51]: so it's assumed that the 2x2 layout is 

wave0       wave2
wave1       wave3

These two assumptions are not about styles, but correctness. Maybe this is the reason why some gemm tests failed in https://github.com/ROCmSoftwarePlatform/triton/pull/402, in which the mfma layout is used directly for global store.

**Review by zhanglx13 — APPROVED** [review]
LGTM.

There might be a disagreement about wave layout that causes the some gemm test failures in https://github.com/ROCmSoftwarePlatform/triton/pull/402. We'll fix that one later.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp (hunk 1 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:51]: You are probably right. I've investigated failures in 402 a little a few weeks ago, and found that current mfma layout is not compatible with global store implementation.

**Pull request merged**